projects/packages/connection/src/sso/class-sso.php

```diff
@@ -0,0 +1,1280 @@
+			'action'       => 'jetpack-sso',
+			'site_id'      => Manager::get_site_id( true ),
+			'sso_nonce'    => $sso_nonce,
+			'reauth'       => '1',
+			'redirect_to'  => rawurlencode( $redirect ),
+			'calypso_auth' => '1',
+		);
+
+		$args = wp_parse_args( $args, $defaults );
+
+		if ( is_wp_error( $args['sso_nonce'] ) ) {
+			return $args['sso_nonce'];
+		}
+
+		return add_query_arg( $args, 'https://wordpress.com/wp-login.php' );
+	}
+
+	/**
+	 * Determines local user associated with a given WordPress.com user ID.
+	 *
+	 * @since jetpack-2.6.0
+	 *
+	 * @param int $wpcom_user_id User ID from WordPress.com.
+	 * @return null|object Local user object if found, null if not.
+	 */
+	public static function get_user_by_wpcom_id( $wpcom_user_id ) {
+		$user_query = new WP_User_Query(
+			array(
+				'meta_key'   => 'wpcom_user_id',
+				'meta_value' => (int) $wpcom_user_id,
+				'number'     => 1,
+			)
+		);
+
+		$users = $user_query->get_results();
+		return $users ? array_shift( $users ) : null;
+	}
+
+	/**
+	 * When jetpack-sso-auth-redirect query parameter is set, will redirect user to
+	 * WordPress.com authorization flow.
+	 *
+	 * We redirect here instead of in handle_login() because Jetpack::init()->build_connect_url
+	 * calls menu_page_url() which doesn't work properly until admin menus are registered.
+	 */
+	public function maybe_authorize_user_after_sso() {
+		$jetpack = Jetpack::init();
+
+		if ( empty( $_GET['jetpack-sso-auth-redirect'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+			return;
+		}
+
+		$redirect_to         = ! empty( $_GET['redirect_to'] ) ? esc_url_raw( wp_unslash( $_GET['redirect_to'] ) ) : admin_url(); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		$request_redirect_to = ! empty( $_GET['request_redirect_to'] ) ? esc_url_raw( wp_unslash( $_GET['request_redirect_to'] ) ) : $redirect_to; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		/** This filter is documented in core/src/wp-login.php */
+		$redirect_after_auth = apply_filters( 'login_redirect', $redirect_to, $request_redirect_to, wp_get_current_user() );
+
+		/**
+		 * Since we are passing this redirect to WordPress.com and therefore can not use wp_safe_redirect(),
+		 * let's sanitize it here to make sure it's safe. If the redirect is not safe, then use admin_url().
+		 */
+		$redirect_after_auth = wp_sanitize_redirect( $redirect_after_auth );
+		$redirect_after_auth = wp_validate_redirect( $redirect_after_auth, admin_url() );
+
+		/**
+		 * Return the raw connect URL with our redirect and attribute connection to SSO.
+		 * We remove any other filters that may be turning on the in-place connection
+		 * since we will be redirecting the user as opposed to iFraming.
+		 */
+		remove_all_filters( 'jetpack_use_iframe_authorization_flow' );
+		add_filter( 'jetpack_use_iframe_authorization_flow', '__return_false' );
+		$connect_url = $jetpack->build_connect_url( true, $redirect_after_auth, 'sso' );
```

Comment:
On the other hand, that would introduce another step into the SSO-connection flow, making it visually longer.
Maybe we'd be better off creating a ["webhook" handler](https://github.com/Automattic/jetpack/blob/2ea4057079bd6a17f282523f80c18b13f1836f9b/projects/packages/connection/src/class-webhooks.php#L85) similar to Jetpack's `action=register`.

Reply:
My mistake, I didn't realize that we don't need site registration here, as SSO will not work on a non-registered site.
You're right, `get_authorization_url()` should work there, let's scratch everything else I said 🙂

Reply:
I'm giving this a try in https://github.com/Automattic/jetpack/pull/36989
I've had to make a few extra changes though, since `get_authorization_url()` alone didn't include all the query strings that come with `Jetpack`'s `build_connect_url`. Let me know what you think about it!